File: debezium/persistent/MongoUpsertHandler.java
package com.tiket.tix.hotel.common.debezium.persistent;

import com.mongodb.client.MongoClient;
import com.mongodb.client.MongoCollection;
import com.mongodb.client.MongoDatabase;
import com.mongodb.client.model.IndexOptions;
import com.mongodb.client.model.Indexes;
import com.mongodb.client.model.ReplaceOptions;
import com.tiket.tix.hotel.common.debezium.config.DebeziumConfiguration;
import com.tiket.tix.hotel.common.debezium.DebeziumHelper;
import java.util.ArrayList;
import org.bson.Document;

public class MongoUpsertHandler implements UpsertHandler {
  private final MongoClient mongoClient;
  private final DebeziumConfiguration config;

  public MongoUpsertHandler(DebeziumConfiguration config, MongoClient mongoClient) {
    this.config = config;
    this.mongoClient = mongoClient;
    migrateMongo(mongoClient);
  }

  @Override
  public void upsert(String json) {
    try {
      Document root = Document.parse(json);
      Document payload = (Document) root.get("payload");
      Document after = payload != null ? (Document) payload.get("after") : null;

      if (after != null) {
        Object id = after.get("id");
        MongoDatabase db = mongoClient.getDatabase(config.getTargetConfiguration().getMongoDatabase());
        MongoCollection<Document> collection = db.getCollection(config.getTargetConfiguration().getMongoCollection());
        collection.replaceOne(
            new Document("_id", id),
            after,
            new ReplaceOptions().upsert(true)
        );
      }
      DebeziumHelper.sendMetric("mongo_upsert_success");
    } catch (Exception e) {
      DebeziumHelper.sendMetric("mongo_upsert_failed");
      e.printStackTrace();
    }
  }


  private void migrateMongo(MongoClient mongoClients) {
    try {
      MongoDatabase db = mongoClients.getDatabase(config.getTargetConfiguration().getMongoDatabase());
      if (!db.listCollectionNames().into(new ArrayList<>())
          .contains(config.getTargetConfiguration().getMongoCollection())) {
        db.createCollection(config.getTargetConfiguration().getMongoCollection());
        db.getCollection(config.getTargetConfiguration().getMongoCollection())
            .createIndex(Indexes.ascending("variable"), new IndexOptions().unique(true));
        System.out.println("Mongo migration done.");
      }
    } catch (Exception e) {
      e.printStackTrace();
    }
  }

}

File: debezium/persistent/mongo_upsert_handler.go
package persistent

import (
	"context"
	"encoding/json"
	"fmt"
	"log"

	"github.com/tiket/TIX-HOTEL-UTILITIES-GO/debezium/config"
	"github.com/tiket/TIX-HOTEL-UTILITIES-GO/debezium/helper"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
)

type MongoUpsertHandler struct {
	Client *mongo.Client
	cfg    *config.DebeziumConfig
	col    *mongo.Collection
}

func NewMongoUpsertHandler(cfg *config.DebeziumConfig) *MongoUpsertHandler {
	client, _ := mongo.Connect(context.Background(), options.Client().ApplyURI(cfg.TargetConfig.MongoURI))
	db := client.Database(cfg.TargetConfig.MongoDB)
	col := db.Collection(cfg.TargetConfig.MongoCol)

	h := &MongoUpsertHandler{
		Client: client,
		cfg:    cfg,
		col:    col,
	}

	h.migrateMongo()
	return h
}

func (h *MongoUpsertHandler) Upsert(jsonStr string) error {
	var root map[string]interface{}
	if err := json.Unmarshal([]byte(jsonStr), &root); err != nil {
		log.Println("Mongo upsert failed: unmarshal error", err)
		return err
	}

	payload, ok := root["payload"].(map[string]interface{})
	if !ok {
		return nil
	}

	after, ok := payload["after"].(map[string]interface{})
	if !ok || after == nil {
		return nil
	}

	id := after["id"]
	ctx := context.Background()

	filter := bson.M{"_id": id}
	opts := options.Replace().SetUpsert(true)

	_, err := h.col.ReplaceOne(ctx, filter, after, opts)
	if err != nil {
		helper.SendMetric(helper.TagBuilder{
			EventType: "mongo_upsert_failed",
		})
		return err
	}

	helper.SendMetric(helper.TagBuilder{
		EventType: "mongo_upsert_success",
	})
	return nil
}

func (h *MongoUpsertHandler) migrateMongo() {
	ctx := context.Background()

	indexModel := mongo.IndexModel{
		Keys:    bson.M{"variable": 1},
		Options: options.Index().SetUnique(true),
	}

	if _, err := h.col.Indexes().CreateOne(ctx, indexModel); err != nil {
		log.Println("Mongo create index failed:", err)
		return
	}

	fmt.Println("Mongo migration done.")
}